File: bindings/python/perf/multi/perf_multi_spot_sendsend_client.py
import sys
import threading
import time

import zlink

from perf_multi_common import (
    TOPIC,
    apply_multi_auto_hwm_msg_unit,
    apply_multi_spot_node_admission,
    benchmark_endpoint,
    bind_spot_node_endpoint,
    benchmark_run_id,
    configure_multi_tls_client,
    configure_multi_tls_server,
    is_active_message,
    latency_ns_from_message,
    parse_client_args,
    perf_client_context,
    print_result_lines,
    publish_control_payload,
    receive_control_payload,
    resolve_multi_connect_ready_timeout_ms,
    resolve_multi_spot_control_settle_s,
    resolve_multi_spot_ready_settle_s,
    result_metrics,
    send_to_spot_nonblocking,
    stamp_payload,
)


SERVER_NODE_RID = b"SPOT-SENDSEND-SERVER-NODE"
SERVER_SPOT_RID = b"SPOT-SENDSEND-SERVER-SPOT"


def _active_slot_limit(total_slots, msg_size):
    if msg_size >= 131072:
        return min(total_slots, 8)
    if msg_size >= 65536:
        return min(total_slots, 32)
    return total_slots


def _drain_reply(spot, *, expected_msg_size, run_id, active_deadline, latencies, record):
    progressed = False
    received = zlink.Received()
    while True:
        try:
            has_received = spot.recv_routed_into(
                received, flags=zlink.RecvFlags.DONT_WAIT
            )
        except zlink.RecvError as exc:
            if exc.result == zlink.RecvResult.NO_DATA:
                return progressed
            raise
        if not has_received:
            return progressed
        progressed = True
        with received:
            if received.request_seq not in (None, 0):
                continue
            parts = received.to_bytes_list()
            if not parts:
                continue
            data = parts[-1]
            if (
                record
                and time.perf_counter() < active_deadline
                and is_active_message(
                    data, expected_msg_size=expected_msg_size, run_id=run_id
                )
            ):
                latency = latency_ns_from_message(data)
                if latency is not None:
                    latencies.append(latency / 2.0)


def main(argv=None):
    args = parse_client_args(argv or sys.argv[1:], pattern="spot_sendsend")
    if not args.control_endpoint:
        raise SystemExit("spot sendsend client expects --endpoint and --control-endpoint")

    run_id = benchmark_run_id()
    latencies = []
    ready_settle_s = resolve_multi_spot_ready_settle_s()
    control_settle_s = resolve_multi_spot_control_settle_s()
    handshake_timeout_s = resolve_multi_connect_ready_timeout_ms() / 1000.0
    runner_start = threading.Event()
    control_connected = threading.Event()

    def stdin_loop():
        for line in sys.stdin:
            text = line.strip()
            if not text:
                continue
            if text == f"START,{args.msg_size}":
                runner_start.set()
            elif text.startswith("CONTROL_CONNECTED,"):
                control_connected.set()
            elif text in {"STOP", "QUIT"}:
                return

    threading.Thread(target=stdin_loop, daemon=True).start()

    with perf_client_context() as ctx:
        apply_multi_auto_hwm_msg_unit(ctx, args.msg_size)
        data_node = zlink.SpotNode(ctx)
        control_node = zlink.SpotNode(ctx)
        configure_multi_tls_server(data_node, args.transport)
        configure_multi_tls_client(data_node, args.transport)
        configure_multi_tls_server(control_node, args.transport)
        configure_multi_tls_client(control_node, args.transport)
        apply_multi_spot_node_admission(data_node, control_node)
        data_node.set_routing_id(b"SPOT-SENDSEND-CLIENT-NODE")
        control_node.set_routing_id(b"SPOT-SENDSEND-CONTROL-CLIENT-NODE")

        control_pub = control_node.create_spot()
        control_sub = control_node.create_spot()
        control_pub.set_routing_id(b"SPOT-SENDSEND-CONTROL-CLIENT-PUB")
        control_sub.set_routing_id(b"SPOT-SENDSEND-CONTROL-CLIENT-SUB")
        control_sub.set_subscription(TOPIC)
        control_endpoint = bind_spot_node_endpoint(
            control_node, args.transport, "multi-spot-sendsend-control-client"
        )
        control_node.connect_peer(args.control_endpoint)
        print(f"CLIENT_CONTROL_ENDPOINT,{control_endpoint}", flush=True)

        data_node.set_router_bind_endpoint(
            benchmark_endpoint(args.transport, "multi-spot-sendsend-client-router")
        )
        data_endpoint_local = bind_spot_node_endpoint(
            data_node, args.transport, "multi-spot-sendsend-client"
        )
        data_node.connect_peer(args.endpoint)
        spots = []
        payloads = [bytearray(args.msg_size) for _ in range(args.clients)]
        seqs = [1 for _ in range(args.clients)]
        waiting = [False for _ in range(args.clients)]
        for index in range(args.clients):
            spot = data_node.create_spot()
            spot.set_routing_id(f"SPOT-SENDSEND-{index}".encode("ascii"))
            spots.append(spot)
        use_poll_wait = args.msg_size < 65536
        poller = None
        poll_events = None
        if use_poll_wait:
            poller = zlink.Poller()
            poll_events = zlink.PollEvents(max(1, len(spots)))
            for index, spot in enumerate(spots):
                poller.add_socket(spot, zlink.PollEventFlag.POLLIN, index)

        if not control_connected.wait(timeout=handshake_timeout_s):
            raise RuntimeError("runner control-connected handshake timeout")

        time.sleep(ready_settle_s)
        time.sleep(control_settle_s)
        if not publish_control_payload(
            control_pub,
            f"DATA_ENDPOINT,{data_endpoint_local}",
            timeout_s=handshake_timeout_s,
        ):
            raise RuntimeError("spot sendsend data endpoint publish timeout")
        time.sleep(control_settle_s)
        publish_control_payload(control_pub, "CONNECTED", timeout_s=handshake_timeout_s)

        probe_deadline = time.perf_counter() + handshake_timeout_s
        probe = stamp_payload(bytearray(args.msg_size), phase=0, run_id=run_id, seq=0)
        probe_waiting = False
        while time.perf_counter() < probe_deadline:
            if not probe_waiting and send_to_spot_nonblocking(
                spots[0], SERVER_NODE_RID, SERVER_SPOT_RID, probe
            ):
                probe_waiting = True
            if _drain_reply(
                spots[0],
                expected_msg_size=args.msg_size,
                run_id=run_id,
                active_deadline=probe_deadline,
                latencies=[],
                record=False,
            ):
                break
            if use_poll_wait:
                poller.wait(poll_events, 1)
            else:
                time.sleep(0.001)
        else:
            raise RuntimeError("spot sendsend probe-ready timeout")

        if not publish_control_payload(
            control_pub,
            f"READY_COUNT,{args.msg_size},{args.clients}",
            timeout_s=handshake_timeout_s,
        ):
            raise RuntimeError("spot sendsend ready publish timeout")
        print(f"CLIENT_READY,{args.msg_size}", flush=True)

        if not runner_start.wait(timeout=handshake_timeout_s):
            raise RuntimeError("spot sendsend runner start handshake timeout")

        direct_start_deadline = time.perf_counter() + handshake_timeout_s
        direct_started = False
        while time.perf_counter() < direct_start_deadline:
            payload_text = receive_control_payload(control_sub)
            if payload_text is None:
                time.sleep(0.001)
                continue
            if payload_text == f"START,{args.msg_size}":
                direct_started = True
                break
        if not direct_started:
            raise RuntimeError("spot sendsend direct start handshake timeout")

        active_slots = _active_slot_limit(len(spots), args.msg_size)
        active_deadline = time.perf_counter() + args.duration
        while time.perf_counter() < active_deadline:
            progressed = False
            for index in range(active_slots):
                spot = spots[index]
                if waiting[index]:
                    if _drain_reply(
                        spot,
                        expected_msg_size=args.msg_size,
                        run_id=run_id,
                        active_deadline=active_deadline,
                        latencies=latencies,
                        record=True,
                    ):
                        waiting[index] = False
                        progressed = True
                    continue
                payload = stamp_payload(
                    payloads[index],
                    phase=1,
                    run_id=run_id,
                    seq=seqs[index],
                )
                if send_to_spot_nonblocking(
                    spot, SERVER_NODE_RID, SERVER_SPOT_RID, payload
                ):
                    waiting[index] = True
                    seqs[index] += 1
                    progressed = True
                else:
                    progressed |= _drain_reply(
                        spot,
                        expected_msg_size=args.msg_size,
                        run_id=run_id,
                        active_deadline=active_deadline,
                        latencies=latencies,
                        record=True,
                    )
            if not progressed:
                remaining_ms = int((active_deadline - time.perf_counter()) * 1000)
                if remaining_ms <= 0:
                    break
                if use_poll_wait:
                    poller.wait(poll_events, min(10, remaining_ms))
                else:
                    time.sleep(0.001)

        drain_deadline = time.perf_counter() + 1.0
        while any(waiting[:active_slots]) and time.perf_counter() < drain_deadline:
            progressed = False
            for index in range(active_slots):
                if not waiting[index]:
                    continue
                if _drain_reply(
                    spots[index],
                    expected_msg_size=args.msg_size,
                    run_id=run_id,
                    active_deadline=active_deadline,
                    latencies=latencies,
                    record=True,
                ):
                    waiting[index] = False
                    progressed = True
            if not progressed:
                if use_poll_wait:
                    poller.wait(poll_events, 20)
                else:
                    time.sleep(0.001)

        if not latencies:
            raise RuntimeError(
                "multi spot sendsend benchmark did not receive any active reply"
            )
        metrics = result_metrics(
            count=len(latencies),
            msg_size=args.msg_size,
            elapsed_s=max(args.duration, 0.001),
            latencies_ns=latencies,
            bandwidth_multiplier=2.0,
        )
        print_result_lines("MULTI_SPOT_SENDSEND", args.transport, args.msg_size, metrics)


if __name__ == "__main__":
    main()

File: bindings/rust/perf/multi/src/perf_multi_spot_sendsend_client.rs
#[path = "perf_common.rs"]
mod common;

use std::io::{self, BufRead, Write};
use std::sync::mpsc;
use std::thread;
use std::time::{Duration, Instant};

use zlink::*;

const PATTERN: &str = "MULTI_SPOT_SENDSEND";
const TOPIC: &str = "bench";
const SERVER_NODE_RID: &[u8] = b"SPOT-SENDSEND-SERVER-NODE";
const SERVER_SPOT_RID: &[u8] = b"SPOT-SENDSEND-SERVER-SPOT";

enum ClientEvent {
    RunnerControlConnected,
    RunnerStart,
    Stop,
}

struct Slot {
    spot: Box<Spot>,
    payload: Vec<u8>,
    seq: u64,
    waiting_reply: bool,
}

fn env_u64(name: &str, default: u64) -> u64 {
    std::env::var(name)
        .ok()
        .and_then(|value| value.parse().ok())
        .unwrap_or(default)
}

fn setup_tls_server(node: &SpotNode, transport: &str) {
    if matches!(transport, "tls" | "wss") {
        let tls = common::resolve_perf_tls_paths().expect("TLS certs not found");
        node.set_tls_server(&tls.cert, &tls.key, false)
            .expect("spot tls server");
    }
}

fn setup_tls_client(node: &SpotNode, transport: &str) {
    if matches!(transport, "tls" | "wss") {
        let tls = common::resolve_perf_tls_paths().expect("TLS certs not found");
        node.set_tls_client(&tls.ca, "localhost", false)
            .expect("spot tls client");
    }
}

fn control_payload(control_sub: &Spot) -> Option<String> {
    let mut message = TopicMessage::empty();
    match control_sub.subscribe(&mut message, RecvFlags::DONT_WAIT) {
        Ok(true) => {
            let data = common::message_payload(message.parts());
            Some(String::from_utf8_lossy(data).into_owned())
        }
        Ok(false) => None,
        Err(err) if err.code() == RecvResult::NoData => None,
        Err(err) => panic!("control subscribe failed: {err}"),
    }
}

fn publish_control(control_pub: &Spot, payload: &str, timeout: Duration) -> bool {
    let deadline = Instant::now() + timeout;
    while Instant::now() < deadline {
        match control_pub
            .publish(TOPIC)
            .message(Message::copy_from(payload.as_bytes()).expect("control message"))
            .flags(SendFlags::DONT_WAIT)
            .submit()
        {
            Ok(_) => return true,
            Err(err) if err.code() == SubmitResult::Backpressured => {
                common::poll_idle_until(deadline, Duration::from_millis(10));
            }
            Err(err) => panic!("control publish failed: {err}"),
        }
    }
    false
}

fn send_payload(spot: &Spot, node_rid: &RoutingId, spot_rid: &RoutingId, payload: &[u8]) -> bool {
    let message = Message::copy_from(payload).expect("payload message");
    match spot
        .send_to_spot(node_rid.clone(), spot_rid.clone())
        .message(message)
        .flags(SendFlags::DONT_WAIT)
        .submit()
    {
        Ok(sent) => sent,
        Err(_) => false,
    }
}

fn active_spot_slot_limit(total: usize, msg_size: usize) -> usize {
    if msg_size >= 131_072 {
        total.min(8)
    } else if msg_size >= 65_536 {
        total.min(24)
    } else {
        total
    }
}

fn drain_slot(
    slot: &mut Slot,
    expected_size: usize,
    deadline: Instant,
    latency: &mut common::LatencyStats,
    record: bool,
) -> bool {
    let mut progressed = false;
    loop {
        let mut received = Received::empty();
        match slot.spot.recv_routed(&mut received, RecvFlags::DONT_WAIT) {
            Ok(true) => {
                progressed = true;
                slot.waiting_reply = false;
                if received.request_seq().unwrap_or(0) != 0 {
                    continue;
                }
                let data = common::message_payload(received.parts());
                if record
                    && Instant::now() <= deadline
                    && common::is_valid_active_message(data, expected_size)
                {
                    let sent_ts_ns = common::decode_sent_ts_ns(data);
                    let now_ns = common::now_ns();
                    if sent_ts_ns > 0 && now_ns >= sent_ts_ns as u64 {
                        latency.record_ns((now_ns - sent_ts_ns as u64) as f64 / 2.0);
                    }
                }
            }
            Ok(false) => break,
            Err(err) if err.code() == RecvResult::NoData => break,
            Err(_) => break,
        }
    }
    progressed
}

fn handle_poll_events(
    events: &[PollEvent],
    slots: &mut [Slot],
    expected_size: usize,
    deadline: Instant,
    latency: &mut common::LatencyStats,
    record: bool,
) -> bool {
    let mut progressed = false;
    for event in events {
        let index = event.slot;
        if index >= slots.len() || event.revents & POLLIN == 0 {
            continue;
        }
        progressed |= drain_slot(&mut slots[index], expected_size, deadline, latency, record);
    }
    progressed
}

fn any_waiting_reply(slots: &[Slot], active_slots: usize) -> bool {
    slots
        .iter()
        .take(active_slots.min(slots.len()))
        .any(|slot| slot.waiting_reply)
}

fn poll_timeout_until(deadline: Instant, max_wait: Duration) -> i64 {
    let remaining = deadline.saturating_duration_since(Instant::now());
    if remaining.is_zero() {
        return 0;
    }
    remaining.min(max_wait).as_millis().max(1) as i64
}

fn main() {
    let args = common::MultiArgs::parse();
    let settings = common::MultiSettings::from_env();
    let (data_endpoint, control_endpoint) = args
        .endpoint
        .split_once(',')
        .expect("spot sendsend client expects data,control endpoint");
    let data_endpoint = data_endpoint.to_string();
    let control_endpoint = control_endpoint.to_string();
    let ready_settle = Duration::from_millis(env_u64("PERF_MULTI_SPOT_READY_SETTLE_MS", 1000));
    let control_settle = Duration::from_millis(env_u64("PERF_MULTI_SPOT_CONTROL_SETTLE_MS", 25));
    let ready_timeout = common::resolve_multi_connect_ready_timeout();
    let (event_tx, event_rx) = mpsc::channel::<ClientEvent>();

    {
        let event_tx = event_tx.clone();
        let msg_size = args.msg_size;
        thread::spawn(move || {
            let stdin = io::stdin();
            for line in stdin.lock().lines() {
                let line = line.unwrap_or_default();
                let text = line.trim();
                if text == format!("START,{msg_size}") {
                    let _ = event_tx.send(ClientEvent::RunnerStart);
                } else if text.starts_with("CONTROL_CONNECTED,") {
                    let _ = event_tx.send(ClientEvent::RunnerControlConnected);
                } else if matches!(text, "STOP" | "QUIT") {
                    let _ = event_tx.send(ClientEvent::Stop);
                    return;
                }
            }
        });
    }

    let ctx = common::perf_client_context();
    common::apply_multi_auto_hwm_msg_unit(&ctx, args.msg_size);
    let data_node = SpotNode::new(&ctx).expect("spot data node");
    let control_node = SpotNode::new(&ctx).expect("spot control node");
    setup_tls_server(&data_node, &args.transport);
    setup_tls_client(&data_node, &args.transport);
    setup_tls_server(&control_node, &args.transport);
    setup_tls_client(&control_node, &args.transport);
    common::apply_multi_spot_node_admission(&data_node, &settings);
    common::apply_multi_spot_node_admission(&control_node, &settings);
    data_node
        .set_routing_id(&RoutingId::from_bytes(b"SPOT-SENDSEND-CLIENT-NODE"))
        .expect("data rid");

    let control_pub = control_node.create_spot().expect("control pub");
    let control_sub = control_node.create_spot().expect("control sub");
    control_sub.set_subscription(TOPIC).expect("control sub");
    let Some(control_bind) = common::benchmark_endpoint(
        PATTERN,
        &args.transport,
        "multi-spot-sendsend-control-client",
    ) else {
        return;
    };
    control_node
        .set_pub_bind(&control_bind)
        .expect("control bind");
    control_node
        .connect_peer(&control_endpoint)
        .expect("connect server control");
    let client_control_endpoint = control_node.last_endpoint().unwrap_or(control_bind);
    println!("CLIENT_CONTROL_ENDPOINT,{client_control_endpoint}");
    io::stdout().flush().ok();

    let Some(data_bind) =
        common::benchmark_endpoint(PATTERN, &args.transport, "multi-spot-sendsend-client")
    else {
        return;
    };
    let Some(data_router_bind) = common::benchmark_endpoint(
        PATTERN,
        &args.transport,
        "multi-spot-sendsend-router-client",
    ) else {
        return;
    };
    data_node
        .set_router_bind(&data_router_bind)
        .expect("client data router bind endpoint");
    data_node
        .set_pub_bind(&data_bind)
        .expect("client data bind");
    let data_endpoint_local = data_node.last_endpoint().unwrap_or(data_bind);
    data_node
        .connect_peer(&data_endpoint)
        .expect("connect data");

    let server_node_rid = RoutingId::from_bytes(SERVER_NODE_RID);
    let server_spot_rid = RoutingId::from_bytes(SERVER_SPOT_RID);
    let mut slots = Vec::with_capacity(settings.clients);
    for index in 0..settings.clients {
        let spot = Box::new(data_node.create_spot().expect("spot"));
        spot.set_routing_id(&RoutingId::from_bytes(
            format!("SPOT-SENDSEND-{index}").as_bytes(),
        ))
        .expect("spot rid");
        slots.push(Slot {
            spot,
            payload: vec![0u8; args.msg_size.max(common::HEADER_SIZE)],
            seq: 1,
            waiting_reply: false,
        });
    }
    let probe_poller = Poller::new().expect("spot sendsend probe poller");
    probe_poller
        .add_socket(&*slots[0].spot, POLLIN, 0)
        .expect("spot sendsend probe poller add");
    let mut probe_events = vec![PollEvent::default(); 1];

    if !matches!(
        event_rx.recv_timeout(ready_timeout),
        Ok(ClientEvent::RunnerControlConnected)
    ) {
        panic!("spot sendsend client control connection timeout");
    }
    thread::sleep(ready_settle);
    thread::sleep(control_settle);
    if !publish_control(
        &control_pub,
        &format!("DATA_ENDPOINT,{data_endpoint_local}"),
        ready_timeout,
    ) {
        panic!("spot sendsend data endpoint publish timeout");
    }
    thread::sleep(control_settle);
    let _ = publish_control(&control_pub, "CONNECTED", ready_timeout);

    let mut probe_payload = vec![0u8; args.msg_size.max(common::HEADER_SIZE)];
    common::encode_header(
        &mut probe_payload,
        common::PHASE_WARMUP,
        args.msg_size as u32,
        0,
    );
    let probe_deadline = Instant::now() + ready_timeout;
    while Instant::now() < probe_deadline {
        if !slots[0].waiting_reply
            && send_payload(
                &slots[0].spot,
                &server_node_rid,
                &server_spot_rid,
                &probe_payload,
            )
        {
            slots[0].waiting_reply = true;
        }
        if drain_slot(
            &mut slots[0],
            args.msg_size,
            probe_deadline,
            &mut common::LatencyStats::new(),
            false,
        ) {
            break;
        }
        common::wait_control_readable_until(&probe_poller, &mut probe_events, probe_deadline);
    }
    if Instant::now() >= probe_deadline {
        panic!("spot sendsend probe-ready timeout");
    }

    if !publish_control(
        &control_pub,
        &format!("READY_COUNT,{},{}", args.msg_size, settings.clients),
        ready_timeout,
    ) {
        panic!("spot sendsend ready publish timeout");
    }
    println!("CLIENT_READY,{}", args.msg_size);
    io::stdout().flush().ok();

    if !matches!(
        event_rx.recv_timeout(ready_timeout),
        Ok(ClientEvent::RunnerStart)
    ) {
        panic!("spot sendsend runner start handshake timeout");
    }
    let control_read_poller = common::control_read_poller(&control_sub);
    let mut control_events = vec![PollEvent::default(); 1];
    let direct_deadline = Instant::now() + ready_timeout;
    let mut direct_started = false;
    while Instant::now() < direct_deadline {
        if let Some(payload) = control_payload(&control_sub) {
            if payload == format!("START,{}", args.msg_size) {
                direct_started = true;
                break;
            }
        }
        common::wait_control_readable_until(
            &control_read_poller,
            &mut control_events,
            direct_deadline,
        );
    }
    if !direct_started {
        panic!("spot sendsend direct start handshake timeout");
    }

    let mut latency = common::LatencyStats::new();
    let deadline = Instant::now() + Duration::from_secs(settings.duration_seconds);
    let active_slots = active_spot_slot_limit(slots.len(), args.msg_size);
    let poller = Poller::new().expect("spot sendsend poller");
    for (index, slot) in slots.iter().enumerate() {
        poller
            .add_socket(&*slot.spot, POLLIN, index)
            .expect("spot sendsend poller add");
    }
    let mut poll_events = vec![PollEvent::default(); slots.len().max(1)];
    while Instant::now() < deadline {
        let mut progressed = false;
        for slot in slots.iter_mut().take(active_slots) {
            if slot.waiting_reply {
                progressed |= drain_slot(slot, args.msg_size, deadline, &mut latency, true);
                continue;
            }
            common::encode_header(
                &mut slot.payload,
                common::PHASE_ACTIVE,
                args.msg_size as u32,
                slot.seq,
            );
            if send_payload(
                &slot.spot,
                &server_node_rid,
                &server_spot_rid,
                &slot.payload,
            ) {
                slot.waiting_reply = true;
                slot.seq += 1;
                progressed = true;
            } else {
                progressed |= drain_slot(slot, args.msg_size, deadline, &mut latency, true);
            }
        }
        if !progressed {
            let wait_ms = poll_timeout_until(deadline, Duration::from_millis(50));
            if wait_ms <= 0 {
                break;
            }
            match poller.wait(&mut poll_events, wait_ms) {
                Ok(event_count) => {
                    handle_poll_events(
                        &poll_events[..event_count],
                        &mut slots,
                        args.msg_size,
                        deadline,
                        &mut latency,
                        true,
                    );
                }
                Err(err) => panic!("spot sendsend poller wait failed: {err}"),
            }
        }
    }

    let pending_deadline = Instant::now() + Duration::from_secs(1);
    while any_waiting_reply(&slots, active_slots) && Instant::now() < pending_deadline {
        let wait_ms = poll_timeout_until(pending_deadline, Duration::from_millis(50));
        if wait_ms <= 0 {
            break;
        }
        match poller.wait(&mut poll_events, wait_ms) {
            Ok(event_count) => {
                handle_poll_events(
                    &poll_events[..event_count],
                    &mut slots,
                    args.msg_size,
                    deadline,
                    &mut latency,
                    true,
                );
            }
            Err(err) => panic!("spot sendsend pending poller wait failed: {err}"),
        }
    }

    common::print_result(
        PATTERN,
        &args.transport,
        args.msg_size,
        settings.duration_seconds,
        &latency.finish(),
    );
}
